File: clients/android/NewsBlur/src/com/newsblur/fragment/LoginRegisterFragment.java
package com.newsblur.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.net.Uri;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.TextView;
import android.widget.TextView.OnEditorActionListener;

import com.newsblur.R;
import com.newsblur.activity.LoginProgress;
import com.newsblur.activity.RegisterProgress;
import com.newsblur.databinding.FragmentLoginregisterBinding;
import com.newsblur.network.APIConstants;
import com.newsblur.util.AppConstants;
import com.newsblur.util.PrefsUtils;

public class LoginRegisterFragment extends Fragment {

	private FragmentLoginregisterBinding binding;

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		final View v = inflater.inflate(R.layout.fragment_loginregister, container, false);
		binding = FragmentLoginregisterBinding.bind(v);

		binding.loginPassword.setOnEditorActionListener(new OnEditorActionListener() {
			@Override
			public boolean onEditorAction(TextView arg0, int actionId, KeyEvent event) {
				if (actionId == EditorInfo.IME_ACTION_DONE ) { 
					logIn();
				}
				return false;
			}
		});

		binding.registrationEmail.setOnEditorActionListener(new OnEditorActionListener() {
			@Override
			public boolean onEditorAction(TextView arg0, int actionId, KeyEvent event) {
				if (actionId == EditorInfo.IME_ACTION_DONE ) { 
					signUp();
				}
				return false;
			}
		});

		return v;
	}

	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		binding.loginButton.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				logIn();
			}
		});
		binding.registrationButton.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				signUp();
			}
		});
		binding.loginChangeToLogin.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				showLogin();
			}
		});
		binding.loginChangeToRegister.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				showRegister();
			}
		});
		binding.loginForgotPassword.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				launchForgotPasswordPage();
			}
		});
		binding.loginCustomServer.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				showCustomServer();
			}
		});
	}

	private void logIn() {
		if (!TextUtils.isEmpty(binding.loginUsername.getText().toString())) {
            // set the custom server endpoint before any API access, even the cookie fetch.
            APIConstants.setCustomServer(binding.loginCustomServerValue.getText().toString());
            PrefsUtils.saveCustomServer(getActivity(), binding.loginCustomServerValue.getText().toString());

			Intent i = new Intent(getActivity(), LoginProgress.class);
			i.putExtra("username", binding.loginUsername.getText().toString());
			i.putExtra("password", binding.loginUsername.getText().toString());
			startActivity(i);
		}
	}

    private void signUp() {
		Intent i = new Intent(getActivity(), RegisterProgress.class);
		i.putExtra("username", binding.registrationUsername.getText().toString());
		i.putExtra("password", binding.registrationPassword.getText().toString());
		i.putExtra("email", binding.registrationEmail.getText().toString());
		startActivity(i);
	}

    private void showLogin() {
        binding.loginViewswitcher.showPrevious();
    }

    private void showRegister() {
        binding.loginViewswitcher.showNext();
    }

    private void launchForgotPasswordPage() {
        try {
            Intent i = new Intent(Intent.ACTION_VIEW);
            i.setData(Uri.parse(AppConstants.FORGOT_PASWORD_URL));
            startActivity(i);
        } catch (Exception e) {
            android.util.Log.wtf(this.getClass().getName(), "device cannot even open URLs to report feedback");
        }
    }

    private void showCustomServer() {
        binding.loginCustomServer.setVisibility(View.GONE);
        binding.loginCustomServerValue.setVisibility(View.VISIBLE);
    }
}

File: clients/android/NewsBlur/app/src/main/java/com/newsblur/fragment/LoginRegisterFragment.kt
package com.newsblur.fragment

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.newsblur.R
import com.newsblur.activity.LoginProgress
import com.newsblur.activity.RegisterProgress
import com.newsblur.databinding.FragmentLoginregisterBinding
import com.newsblur.network.APIConstants
import com.newsblur.util.AppConstants
import com.newsblur.util.PrefsUtils

class LoginRegisterFragment : Fragment() {

    private lateinit var binding: FragmentLoginregisterBinding

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.fragment_loginregister, container, false)
        binding = FragmentLoginregisterBinding.bind(v)

        binding.loginPassword.setOnEditorActionListener { _, actionId: Int, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                logIn()
            }
            false
        }
        binding.registrationEmail.setOnEditorActionListener { _, actionId: Int, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                signUp()
            }
            false
        }
        return v
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.loginButton.setOnClickListener { logIn() }
        binding.registrationButton.setOnClickListener { signUp() }
        binding.loginChangeToLogin.setOnClickListener { showLogin() }
        binding.loginChangeToRegister.setOnClickListener { showRegister() }
        binding.loginForgotPassword.setOnClickListener { launchForgotPasswordPage() }
        binding.loginCustomServer.setOnClickListener { showCustomServer(true) }
        binding.buttonResetUrl.setOnClickListener { showCustomServer(false) }
        val customServerUrl = PrefsUtils.getCustomSever(requireContext())
        if (!TextUtils.isEmpty(customServerUrl)) {
            binding.loginCustomServerValue.setText(customServerUrl)
            showCustomServer(true)
        }
    }

    private fun logIn() {
        if (!TextUtils.isEmpty(binding.loginUsername.text.toString())) {
            // set the custom server endpoint before any API access, even the cookie fetch.
            val customServerValue = binding.loginCustomServerValue.text.toString()
            if (!TextUtils.isEmpty(customServerValue) && customServerValue.startsWith("https://")) {
                APIConstants.setCustomServer(customServerValue)
                PrefsUtils.saveCustomServer(activity, customServerValue)
            } else if (!TextUtils.isEmpty(customServerValue)) {
                Toast.makeText(requireActivity(), R.string.login_custom_server_scheme_error, Toast.LENGTH_LONG).show()
                return
            }

            val loginIntent = Intent(activity, LoginProgress::class.java).apply {
                putExtra("username", binding.loginUsername.text.toString())
                putExtra("password", binding.loginPassword.text.toString())
            }
            startActivity(loginIntent)
        }
    }

    private fun signUp() {
        val registerIntent = Intent(activity, RegisterProgress::class.java).apply {
            putExtra("username", binding.registrationUsername.text.toString())
            putExtra("password", binding.registrationPassword.text.toString())
            putExtra("email", binding.registrationEmail.text.toString())
        }
        startActivity(registerIntent)
    }

    private fun showLogin() {
        binding.loginViewswitcher.showPrevious()
    }

    private fun showRegister() {
        binding.loginViewswitcher.showNext()
    }

    private fun launchForgotPasswordPage() {
        try {
            val i = Intent(Intent.ACTION_VIEW)
            i.data = Uri.parse(AppConstants.FORGOT_PASWORD_URL)
            startActivity(i)
        } catch (e: Exception) {
            Log.wtf(this.javaClass.name, "device cannot even open URLs to report feedback")
        }
    }

    private fun showCustomServer(isVisible: Boolean) {
        binding.loginCustomServer.visibility = if (isVisible) View.GONE else View.VISIBLE
        binding.loginCustomServerValue.visibility = if (isVisible) View.VISIBLE else View.GONE
        binding.buttonResetUrl.visibility = if (isVisible) View.VISIBLE else View.GONE
        binding.textCustomServer.visibility = if (isVisible) View.VISIBLE else View.GONE

        if (!isVisible) {
            binding.loginCustomServerValue.setText("")
            APIConstants.unsetCustomServer()
            PrefsUtils.clearCustomServer(context)
        }
    }
}